### PR TITLE
 SAGE-649: node registration with beekeeper db 

### PR DESCRIPTION
**Pull request by wgerlach**

### CODE REVIEW

File: bk-config/unit-tests/test_register.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+
+set -e
+
+./create_client_files.sh bk-sshd 22 forever
+
+
+# TODO: The known_hosts file did not work here for some reason
+
+NEW_ID=$(docker run --rm  --network beekeeper_default -v `pwd`:/keys/:ro -w /keys -v beekeeper-config_bk-secrets:/usr/lib/sage/:ro sagecontinuum/bk-config bash -c 'apt-get install -y jq ; ssh -o StrictHostKeyChecking=no  sage_registration@bk-sshd -p 22 -i register.pem register 0000000000000001 | jq -r -j .id' | tail -n 1)
```
Comment (jswantek) [inline bk-config/unit-tests/test_register.sh:16]: do we want to define a network in the docker-compose ?

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 2]
```diff
@@ -33,6 +34,7 @@
 CA_FILE = os.path.join(BASE_KEY_DIR, "certca/sage_beekeeper_ca")
 USER_SERVER = "http://bk-sshd"
 
+BEEKEEPER_DB_API = "http://bk-nodes:5000"
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:37]: just checking .. you reference port 5000 here.  in the docker-compose you expose 5000 (for debug) which I assume you will remove at some point.

in order for this to work does bk-register and bk-nodes need to be in the same docker network?

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 3]
```diff
@@ -46,6 +48,15 @@ def register():
     """
     id = flask.request.args.get("id", type=str)
 
+    # check if BEEKEEPER_DB_API is alive
+    try:
+        bk_api_result = requests.get(f'{BEEKEEPER_DB_API}').content
+    except Exception as e:
+        return f"Error: X Beekeeper DB API ({BEEKEEPER_DB_API}) cannot be reached: {str(e)}"
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:55]: i dont exactly remember how this works but what happens if you don't include an `error` code in your `return`?  Is the return code "success"?

I was expecting to see something like `return f"Error...", 500`

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 4]
```diff
@@ -46,6 +48,15 @@ def register():
...
+
+    if 'SAGE Beekeeper' != bk_api_result.decode("utf-8").strip():
+        return "Error: Y Beekeeper DB API ({BEEKEEPER_DB_API}) cannot be reached: \""+bk_api_result.decode("utf-8")+"\""
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:58]: same question (as line 55) about the `return` error code

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 3 above)
Comment (jswantek) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:55]: Also I assume you meant to put a number in place of "X" in the string "Error: X Beekeeper DB"

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 4 above)
Comment (jswantek) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:58]: Also I assume you meant to put a number in place of "Y" in the string "Error: Y Beekeeper DB"

Comment (jswantek) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:58]: Also i think you could have used the same `f` format for your string

```
return f"Error: Y Beekeeper DB API ({BEEKEEPER_DB_API}) cannot be reached: \"{bk_api_result.decode("utf-8")}\""
```

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 5]
```diff
@@ -77,6 +88,20 @@ def register():
         logger.error(e)
         return "Error: unable to register id [{}]\n".format(id), 500
 
+    # update beekeeper db
+    payload = {"node_id": id, "source": "beekeeper-register", "operation":"insert", "field_name": "registration_event", "field_value": datetime.datetime.now().replace(microsecond=0).isoformat()}
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:92]: I am curious.  Is there a way for the SQL table be setup to restrict the accuracy of the `datetime` to be at the seconds level?  So, you don't have to remember to set the microseconds to 0 everywhere?

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 6]
```diff
@@ -77,6 +88,20 @@ def register():
...
+
+    try:
+        bk_api_response = requests.post(f'{BEEKEEPER_DB_API}/log',data=json.dumps(payload))
+    except Exception as e:
+        return f"Error: X Beekeeper DB API ({BEEKEEPER_DB_API}) cannot be reached: {str(e)}"
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:97]: same question (as line 55) about the return error code

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py [hunk 7]
```diff
@@ -77,6 +88,20 @@ def register():
...
+
+    if bk_api_response.status_code != 200 :
+        con  = bk_api_response.content
+        con_str = con.decode("utf-8")
+        return f"Error: Submission to beekeeper DB failed: { con_str}"
```
Comment (jswantek) [inline bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:102]: same question (as line 55) about the return error code

Comment (jswantek) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:102]: I think the extra :space: in `{ con_str}` will present you issues.

File: bk-nodes/bk_nodes.py [hunk 8]
```diff
@@ -63,49 +63,60 @@ def post(self):
...
-                except Exception as ex:
-                    raise ErrorResponse(f"Unexpected error: {e}" , status_code=HTTPStatus.INTERNAL_SERVER_ERROR)
-
-                logData.append(newLogDataEntry)
-                #print("success", flush=True)
-                
 
-            bee_db.nodes_log_add(logData) #  effective_time=effective_time)
+        logData = []
+        default_effective_time = datetime.datetime.now(datetime.timezone.utc) # this way all operations in this submission have the exact same time
```
Comment (jswantek) [inline bk-nodes/bk_nodes.py:86]: i assume you dont want to set the microseconds to 0 here?

File: bk-config/unit-tests/test_register.sh (hunk 1 above)
Comment (wgerlach) [reply bk-config/unit-tests/test_register.sh:16]: Given that we may want to migrate this to k8s anyway, I do not think this is critical. But in general, you are right.

File: bk-nodes/bk_db.py [hunk 9]
```diff
@@ -208,11 +208,17 @@ def getNodeState(self, node_id, timestamp=None):
...
         row = self.cur.fetchone()
+        if not row:
+            raise Exception("Node not found")
+
 
         result = {}
         for i in range(len(row)):
             result[fields[i]] = row[i]
 
+        if not result:
```
[the remark this answers is not in the document]
Comment (wgerlach) [reply bk-nodes/bk_db.py:219]: You are right, that should not happen. That probably is an artifact of how I wrote the code. While I would not intentionally write such a second test, I also won't bother removing it now. If I ever make changes upstream, that check might come in handy.

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 3 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: Good point, I added the error code! I also replaced X and Y with better error descriptions.

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 4 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: fixed

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 5 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py:93]: also good point, I now use TIMESTAMP(0) to disable fractional seconds in the schema

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 6 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: done

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 7 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: done

File: bk-nodes/bk_db.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wgerlach) [reply bk-nodes/bk_db.py:219]: I changed the message (yes, should not happen)

File: bk-register/ROOTFS/opt/sage/beekeeper/register/register.py (hunk 2 above)
Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: I introduced an env variable BEEKEEPER_DB_API . Should we do the same for USER_SERVER ?

Comment (wgerlach) [reply bk-register/ROOTFS/opt/sage/beekeeper/register/register.py]: Port 5000 was only open for debugging purposes. I renamed USER_SERVER and made it configurable via environment variable. Now bk-register and bk-nodes do not need to be inside of a docker network, but bk-register needs to reach bk-nodes somehow.

**Pull request merged**